spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/spanner/GcpSpannerEmulatorAutoConfiguration.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.spanner;
+
+import com.google.cloud.NoCredentials;
+import com.google.cloud.spanner.SpannerOptions;
+
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Provides auto-configuration to use the Spanner emulator if enabled.
+ *
+ * @author Eddú Meléndez
+ */
+@Configuration
+@AutoConfigureBefore(GcpSpannerAutoConfiguration.class)
+@EnableConfigurationProperties(GcpSpannerProperties.class)
+@ConditionalOnProperty(prefix = "spring.cloud.gcp.spanner.emulator", name = "enabled", havingValue = "true")
+public class GcpSpannerEmulatorAutoConfiguration {
+
+	private final GcpSpannerProperties properties;
+
+	public GcpSpannerEmulatorAutoConfiguration(GcpSpannerProperties properties) {
+		this.properties = properties;
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public SpannerOptions spannerOptions() {
+		SpannerOptions.Builder builder = SpannerOptions.newBuilder()
+				.setProjectId(this.properties.getProjectId())
+				.setCredentials(NoCredentials.getInstance());
+		if (this.properties.getEmulatorHost() != null) {
```

Comment:
the client library resolves the `SPANNER_EMULATOR_HOST` environment variable. But, I agree it can be dismiss. I think we can validate if the environment variable or the property is set, otherwise we can fail and say `neither environment variable or property was set`. WDYT?

Reply:
Since we're providing `NoCredentials`, there is no risk of connecting to the real Spanner and incurring costs.

Reply:
In regard to  ".enabled vs the property" - .enabled is the way to go.

Reply:
Yes, let's fail if "enabled" is set but "emulatorHost" is missing, since there is no way to make that configuration work. `Assert.notEmpty` should work.